cmd/thanos/rule.go

```diff
@@ -472,49 +472,26 @@ func runRule(
 	}
 
 	// Handle reload and termination interrupts.
-	reload := make(chan struct{}, 1)
 	{
 		cancel := make(chan struct{})
-		reload <- struct{}{} // Initial reload.
-
 		g.Add(func() error {
+			// Initialize rules.
+			if err := reloadRules(logger, ruleFiles, ruleMgr, evalInterval, configSuccess, configSuccessTime, rulesLoaded); err != nil {
+				level.Error(logger).Log("msg", "initialize rules failed", "err", err)
+				// Return when initialize with invalid pattern error.
+				if _, ok := err.(*errInvalidPattern); ok {
+					return err
```

Comment:
This changes behaviour compared to the previous implementation. Previously, it was logging and proceeding to other rule files.
```go
if err != nil {	
	// The only error can be a bad pattern.	
	level.Error(logger).Log("msg", "retrieving rule files failed. Ignoring file.", "pattern", pat, "err", err)	
	continue	
}
```

If this is what we want, you should update the CHANGELOG accordingly.
You should update CHANGELOG anyway because you are changing the behaviour of the HTTP API.

@bwplotka

Reply:
@kakkoyun  This comes from orignal logic. I wrote it to keep same logic with orignal one.

Reply:
No actually. That's what I'm trying to tell, it doesn't keep the same logic _if I'm not mistaken_. If you check your `reloadRules` function, in a for-loop when execution encounters an error, it returns it so it breaks out from with the first encounter of an error.
```go
var files []string
for _, pat := range ruleFiles {
    fs, err := filepath.Glob(pat)
    if err != nil {
        // Check errInvalidPattern when initialize.
        return &errInvalidPattern{err, pat}
    }

    files = append(files, fs...)
}
```

The old code, just prints an error and continue with the execution and checks other files.

Reply:
@kakkoyun Sorry for my misunderstanding. Cannot remeber why change like this, since it's long time.
Maybe the change is reasonable.  What do you thank?

Reply:
@kakkoyun Hi, the code comes from prometheus. Please have a look at the link 
https://github.com/prometheus/prometheus/blob/fdbb18b445285b69d005d88d2c1edb83f5c1b5e5/cmd/prometheus/main.go#L459.

Reply:
The change is definitely reasonable, all I'm saying is that we should document the behaviour change. And of course, if also the maintainers agree.

Reply:
From my side, I think @kakkoyun is right.

> You should update CHANGELOG anyway because you are changing the behaviour of the HTTP API.

Yes

> The old code, just prints an error and continue with the execution and checks other files.

My opinion is that we should fail and proceed to another rule. No point in failing everything and stop as it does not crash (and should not!) ruler. Also we should be consistent to Prometheus behaviour. (: Can we change the logic here to what was previously? @00arthur00

Reply:
@bwplotka Thanks for your reply. I will revert the logic. I.e, keep log and no termination.

Reply:
@bwplotka I think we should error and exit out if rule file includes faulty rules **when we are starting component for the first time**. And then for any subsequent update, it should just log the error and continue running without any crash. What do you think?